File: firmware/lightweaver-controller/tests/wifi-channel-policy.cpp
// Host coverage for the channel decisions behind the setup hotspot.
//
// The radio behaviour itself is not host-testable — nothing here can prove that
// a real ESP32-S3 stops deauthenticating phones — but the two decisions that
// drive it are pure and are exactly where the last regression lived: a lookup
// that could only ever answer "unknown" on the boot path, so the alignment it
// fed never ran. These assertions pin that it now has a second source.
#include <cassert>
#include <cstdint>

#include "../src/LightweaverWifiChannelPolicy.h"

using lightweaver::kWifiChannelUnknown;
using lightweaver::normalizeWifiChannel;
using lightweaver::planWifiProvenRecord;
using lightweaver::setupApChannel;
using lightweaver::WifiProvenRecord;

static_assert(kWifiChannelUnknown == 0, "unknown must stay distinguishable from every real channel");
static_assert(normalizeWifiChannel(1) == 1, "channel 1 is usable");
static_assert(normalizeWifiChannel(14) == 14, "channel 14 is usable");
static_assert(normalizeWifiChannel(0) == kWifiChannelUnknown, "0 already means unknown");
static_assert(normalizeWifiChannel(15) == kWifiChannelUnknown, "above 2.4GHz is not usable");
static_assert(normalizeWifiChannel(36) == kWifiChannelUnknown, "a 5GHz association cannot carry the soft AP");
static_assert(normalizeWifiChannel(-1) == kWifiChannelUnknown, "an SDK error return must not reach softAP()");

// Fresh scan results beat a remembered channel: during commissioning the setup
// page has just scanned, so they describe the network as it is right now.
static_assert(setupApChannel(6, 11) == 6, "scan results win when present");
static_assert(setupApChannel(6, kWifiChannelUnknown) == 6, "scan results are enough on their own");

// The boot path. The scan cache is always empty there — the speculative boot
// scan was removed because it parked the radio off-channel — so without the
// remembered channel this answers "unknown" and the AP goes up wherever the SDK
// puts it, which is the bug this whole change exists to fix.
static_assert(setupApChannel(kWifiChannelUnknown, 11) == 11,
              "a remembered channel must carry the boot path, where no scan has run");

// A blank card has neither, and must behave exactly as it always has: the AP
// starts on the SDK default and nothing tries to move it.
static_assert(setupApChannel(kWifiChannelUnknown, kWifiChannelUnknown) == kWifiChannelUnknown,
              "a card that has never joined anything must not be aimed at a channel");

// A stored value that is not a real channel is worth no more than none at all.
static_assert(setupApChannel(kWifiChannelUnknown, 99) == kWifiChannelUnknown,
              "a corrupt stored channel must normalize away, not reach softAP()");
static_assert(setupApChannel(200, 11) == 11,
              "a nonsense scan reading falls through to the remembered channel");

int main() {
  // First association on freshly saved credentials: nothing is stored yet, so
  // this is the write that makes the card resumable and remembers the channel.
  WifiProvenRecord first = planWifiProvenRecord(true, false, kWifiChannelUnknown, 6);
  assert(first.persist);
  assert(first.channel == 6);

  // Steady state. This runs on every connectivity poll, so it has to stop
  // writing once there is nothing new to say, or it burns NVS every 250ms.
  WifiProvenRecord steady = planWifiProvenRecord(true, true, 6, 6);
  assert(!steady.persist);

  // The router moved channels since the last boot. The AP went up on the stale
  // channel and the SDK migrated it — no worse than before this change — but
  // the new channel gets written back so the *next* boot is right.
  WifiProvenRecord moved = planWifiProvenRecord(true, true, 6, 11);
  assert(moved.persist);
  assert(moved.channel == 11);

  // Associated somewhere the soft AP could not have followed. A good 2.4GHz
  // memory is still worth more than nothing, so it must survive.
  WifiProvenRecord offBand = planWifiProvenRecord(true, true, 6, 36);
  assert(!offBand.persist);
  assert(offBand.channel == 6);

  // Proven but with no channel yet — a card upgraded from firmware that had no
  // channel field. The first association after the upgrade fills it in.
  WifiProvenRecord backfill = planWifiProvenRecord(true, true, kWifiChannelUnknown, 11);
  assert(backfill.persist);
  assert(backfill.channel == 11);

  // Proven, no stored channel, and nothing usable observed: still nothing to
  // write. An older card on 5GHz must not rewrite its credentials every poll.
  WifiProvenRecord nothingToSay = planWifiProvenRecord(true, true, kWifiChannelUnknown, 36);
  assert(!nothingToSay.persist);
  assert(nothingToSay.channel == kWifiChannelUnknown);

  // No credentials at all: there is no network to attribute a channel to, so
  // nothing is ever persisted. The blank-card path stays untouched.
  WifiProvenRecord blank = planWifiProvenRecord(false, false, kWifiChannelUnknown, 6);
  assert(!blank.persist);

  return 0;
}

File: firmware/lightweaver-controller/tests/wifi-channel-policy.mjs
// The setup hotspot and the station link share one radio. If they are on
// different channels the SDK drags the AP onto the station's channel during the
// join and deauthenticates every phone connected to it. Aligning the AP first
// was already shipped — but its channel lookup read only the scan cache, and
// the boot scan that filled that cache had been removed in the same commit for
// parking the radio off-channel. So the lookup answered "unknown" on every
// boot, the alignment never ran, and a reset card's hotspot could vanish the
// moment it reached the LAN.
//
// The fix is to remember the channel instead of looking for it. These
// assertions pin the pure decisions (compiled and run on the host) and the
// wiring that has to stay in place around them, including that no scan crept
// back onto the AP-start path.
import assert from 'node:assert/strict';
import { execFileSync } from 'node:child_process';
import { mkdtempSync, readFileSync, rmSync } from 'node:fs';
import { tmpdir } from 'node:os';
import { join, resolve } from 'node:path';

const root = resolve(import.meta.dirname, '..');

const dir = mkdtempSync(join(tmpdir(), 'lw-wifi-channel-'));
try {
  const binary = join(dir, 'wifi-channel-policy');
  execFileSync(process.env.CXX || 'c++', [
    '-std=c++17', '-Wall', '-Wextra', '-Werror',
    'wifi-channel-policy.cpp',
    '-o', binary,
  ], { cwd: new URL('.', import.meta.url), stdio: 'inherit' });
  execFileSync(binary, { stdio: 'inherit' });
} finally {
  rmSync(dir, { recursive: true, force: true });
}

const web = readFileSync(resolve(root, 'src/LightweaverWeb.cpp'), 'utf8');
const storage = readFileSync(resolve(root, 'src/LightweaverStorage.cpp'), 'utf8');
const types = readFileSync(resolve(root, 'src/LightweaverTypes.h'), 'utf8');

function functionBody(source, signature) {
  const match = source.match(signature);
  assert.ok(match, `missing function matching ${signature}`);
  const start = match.index;
  const open = source.indexOf('{', start);
  let depth = 0;
  for (let index = open; index < source.length; index += 1) {
    if (source[index] === '{') depth += 1;
    if (source[index] === '}') depth -= 1;
    if (depth === 0) return source.slice(start, index + 1);
  }
  throw new Error(`unterminated body for ${signature}`);
}

// The channel has to live with the credentials, not in runtime state: it is
// only useful across a reboot.
assert.match(types, /struct\s+WifiConfig\s*\{[\s\S]*?uint8_t\s+channel\s*=\s*0;[\s\S]*?\};/,
  'WifiConfig must carry the remembered channel, since it is only useful after a reboot');

// Every path that raises the setup AP has to consult both sources. Reading the
// scan cache alone is the exact regression this replaces.
for (const signature of [
  /void\s+startApMode\s*\(\s*RuntimeConfig&\s*config\s*\)\s*\{/,
  /void\s+ensureRecoveryAp\s*\(\s*RuntimeConfig&\s*config\s*\)\s*\{/,
  /void\s+alignSetupApChannel\s*\(\s*RuntimeConfig&\s*config\s*\)\s*\{/,
]) {
  const body = functionBody(web, signature);
  assert.match(body, /setupApChannelFor\(config\)/,
    `${signature} must pick its channel from both the scan cache and the remembered channel`);
  assert.ok(!/lastScanChannelForSsid/.test(body),
    `${signature} must not read the scan cache directly: it is always empty at boot, which is what made the alignment dead code`);
}

const chooser = functionBody(web, /uint8_t\s+setupApChannelFor\s*\(/);
assert.match(chooser, /lightweaver::setupApChannel\(/,
  'the channel choice must go through the shared policy, which is host-tested');
assert.match(chooser, /config\.wifi\.channel/,
  'the remembered channel is the only source available on the boot path and must be consulted there');
assert.ok(!/WiFi\.scanNetworks/.test(chooser),
  'choosing a channel must never start a scan: scanning is what parks the radio off-channel and knocks phones off the hotspot');

// The regression guard from the scan work, restated for the path this change
// touches: nothing here may re-arm scanning at boot.
const setupWeb = functionBody(web, /void\s+setupLightweaverWeb\s*\(/);
assert.ok(!/WiFi\.scanNetworks/.test(setupWeb),
  'boot must not start a scan to discover the channel; that is what the persisted channel exists to avoid');

// The channel is free at association time — the radio is sitting on it.
const maintain = functionBody(web, /void\s+maintainConnectivity\s*\(\s*\)\s*\{/);
assert.match(maintain, /markWifiCredentialsProven\(cfg,\s*associatedStationChannel\(\)\)/,
  'a successful association must record the channel it landed on, which is the only way a later boot knows it without scanning');
assert.match(functionBody(web, /uint8_t\s+associatedStationChannel\s*\(/), /WiFi\.channel\(\)/,
  'the association channel comes from the radio, not from a scan');

// Storage round-trip: written on association, read back at boot, cleared when
// the credentials it describes are replaced.
const proven = functionBody(storage, /bool\s+markWifiCredentialsProven\s*\(/);
assert.match(proven, /lightweaver::planWifiProvenRecord\(/,
  'when to persist is a decision with several cases and belongs in the host-tested policy');
assert.match(proven, /out\["channel"\]\s*=\s*record\.channel/,
  'the channel must be written into the stored credential blob');

const overlay = functionBody(storage, /void\s+overlayNvsWifi\s*\(/);
assert.match(overlay, /config\.wifi\.channel\s*=\s*lightweaver::normalizeWifiChannel\(doc\["channel"\]/,
  'the stored channel must be read back at boot and clamped, so a corrupt blob cannot reach WiFi.softAP()');

const save = functionBody(storage, /bool\s+saveWifiConfigJson\s*\(/);
assert.match(save, /out\["channel"\]\s*=\s*0/,
  'new credentials must clear the remembered channel: it described the previous network and would aim the hotspot at the wrong channel');

assert.match(functionBody(storage, /void\s+resetWifi\s*\(/), /wifi\.channel\s*=\s*0/,
  'clearing the credentials must clear the channel that described them');

console.log('wifi-channel-policy tests passed');
